// vite_react_shadcn_ts/src/lib/apiClient.ts
// Centralized API Client - single source of truth for all backend communication
// Handles auth tokens, error normalization, and CORS-safe requests

import { API_BASE_URL } from './apiConfig';
import type {
  ArtistPermissions,
  ArtistTag,
  CompanyProfile,
  CompanyProfileUpdate,
  ScanResponse,
  ComplianceEvent,
  ScanResult,
} from './apiTypes';

export interface ApiErrorDetails {
  message: string;
  status: number;
  code?: string;
  raw?: unknown;
}

export class ApiError extends Error {
  status: number;
  code?: string;
  raw?: unknown;

  constructor(details: ApiErrorDetails) {
    super(details.message);
    this.name = 'ApiError';
    this.status = details.status;
    this.code = details.code;
    this.raw = details.raw;
  }

  static fromResponse(status: number, data: unknown): ApiError {
    let message = `Request failed with status ${status}`;
    let code: string | undefined;

    if (data && typeof data === 'object') {
      const errorData = data as Record<string, unknown>;
      if (errorData.error && typeof errorData.error === 'object') {
        const err = errorData.error as Record<string, unknown>;
        message = (err.message as string) || message;
        code = err.code as string;
      } else if (errorData.message) {
        message = errorData.message as string;
      }
    }

    return new ApiError({ message, status, code, raw: data });
  }

  static fromNetworkError(error: Error): ApiError {
    let message = error.message || 'Network request failed';
    
    // Detect CORS errors
    if (error.message.includes('Failed to fetch') || error.message.includes('NetworkError')) {
      message = 'Network error - possibly CORS blocked or server unreachable';
    }
    
    return new ApiError({ message, status: 0, code: 'NETWORK_ERROR', raw: error });
  }
}

// ==================== Request Types ====================

interface RequestOptions {
  headers?: Record<string, string>;
  skipAuth?: boolean;
}

// ==================== API Client Class ====================

class ApiClient {
  private baseUrl: string;
  private accessToken: string | null = null;

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl.replace(/\/$/, ''); // Remove trailing slash
  }

  // ==================== Token Management ====================

  setAccessToken(token: string | null): void {
    this.accessToken = token;
  }

  getAccessToken(): string | null {
    return this.accessToken;
  }

  hasToken(): boolean {
    return this.accessToken !== null;
  }

  // ==================== Core Request Method ====================

  private async request<T>(
    method: string,
    endpoint: string,
    body?: unknown,
    options: RequestOptions = {}
  ): Promise<T> {
    const { headers: customHeaders = {}, skipAuth = false } = options;

    const headers: Record<string, string> = { ...customHeaders };

    // Add Authorization header if token exists and auth not skipped
    if (!skipAuth && this.accessToken) {
      headers['Authorization'] = `Bearer ${this.accessToken}`;
    }

    // Add Content-Type for JSON requests (not for FormData)
    if (body && !(body instanceof FormData)) {
      headers['Content-Type'] = 'application/json';
    }

    const url = `${this.baseUrl}${endpoint}`;

    try {
      const response = await fetch(url, {
        method,
        headers,
        body: body instanceof FormData ? body : body ? JSON.stringify(body) : undefined,
      });

      // Parse response
      const text = await response.text();
      let data: unknown;
      
      try {
        data = text ? JSON.parse(text) : {};
      } catch {
        data = { raw: text };
      }

      if (!response.ok) {
        throw ApiError.fromResponse(response.status, data);
      }

      return data as T;
    } catch (error) {
      if (error instanceof ApiError) {
        throw error;
      }
      throw ApiError.fromNetworkError(error as Error);
    }
  }

  // ==================== HTTP Methods ====================

  async get<T>(endpoint: string, options?: RequestOptions): Promise<T> {
    return this.request<T>('GET', endpoint, undefined, options);
  }

  async post<T>(endpoint: string, body?: unknown, options?: RequestOptions): Promise<T> {
    return this.request<T>('POST', endpoint, body, options);
  }

  async put<T>(endpoint: string, body?: unknown, options?: RequestOptions): Promise<T> {
    return this.request<T>('PUT', endpoint, body, options);
  }

  async delete<T>(endpoint: string, options?: RequestOptions): Promise<T> {
    return this.request<T>('DELETE', endpoint, undefined, options);
  }

  // ==================== File Upload Helper ====================

  async upload<T>(
    endpoint: string,
    files: File[],
    fieldName: string = 'files[]',
    additionalData?: Record<string, unknown>,
    options?: RequestOptions
  ): Promise<T> {
    const formData = new FormData();

    files.forEach((file) => {
      formData.append(fieldName, file);
    });

    if (additionalData) {
      Object.entries(additionalData).forEach(([key, value]) => {
        if (value === undefined || value === null) {
          return;
        }

        if (value instanceof File) {
          formData.append(key, value);
          return;
        }

        if (Array.isArray(value) || typeof value === 'object') {
          formData.append(key, JSON.stringify(value));
          return;
        }

        formData.append(key, String(value));
      });
    }

    // Don't set Content-Type for FormData - browser handles multipart boundary
    return this.request<T>('POST', endpoint, formData, options);
  }

  // ==================== Health Check ====================

  async checkHealth(): Promise<{ ok: boolean; data?: Record<string, unknown> }> {
    return this.get('/health', { skipAuth: true });
  }

  // ==================== Auth Endpoints ====================

  async getAuthMe(): Promise<{ ok: boolean; data: { sub: string; role?: string; [key: string]: unknown } }> {
    const response = await this.get<{ ok: boolean; data: { sub?: string; auth0UserId?: string; role?: string; [key: string]: unknown } }>('/api/v1/auth/me');
    return {
      ok: response.ok,
      data: {
        ...response.data,
        sub: response.data?.sub || response.data?.auth0UserId || 'unknown'
      }
    };
  }

  async getMe(): Promise<{ ok: boolean; user: { id: string; auth0UserId: string; role: string | null } }> {
    return this.get('/api/v1/me');
  }

  async setMyRole(role: 'artist' | 'company'): Promise<{ ok: boolean; data: { id: string; auth0UserId: string; role: string } }> {
    return this.put('/api/v1/me/role', { role });
  }

  // ==================== Artist Endpoints ====================

  async checkArtistAccess(): Promise<{ ok: boolean; data: Record<string, unknown> }> {
    return this.get('/api/v1/artist/stub');
  }

  async getArtistTags(): Promise<ArtistTag[]> {
    const response = await this.get<{ ok: boolean; data: ArtistTag[] }>('/api/v1/artist/tags');
    return response.data || [];
  }

  async uploadArtwork(files: File[], permissions: ArtistPermissions): Promise<void> {
    const allowedUseCases = [...permissions.allowed_use_cases];
    const otherUseCase = permissions.other_use_case?.trim();
    if (otherUseCase) {
      allowedUseCases.push(otherUseCase);
    }

    const payload = {
      ai_training_allowed: permissions.ai_training,
      allowed_use_cases: allowedUseCases,
      attribution_required: permissions.attribution,
      notes: permissions.notes || ''
    };

    if (files.length === 1) {
      // Single upload: POST /api/v1/artist/artworks/upload, field name "file"
      await this.upload(
        '/api/v1/artist/artworks/upload',
        files,
        'file',
        payload
      );
      return;
    }
    // Batch upload: POST /api/v1/artist/artworks/upload-batch, field name "files"
    await this.upload(
      '/api/v1/artist/artworks/upload-batch',
      files,
      'files',
      payload
    );
  }

  async updateTagPermissions(tagId: string, permissions: ArtistPermissions): Promise<ArtistTag> {
    const allowedUseCases = [...permissions.allowed_use_cases];
    const otherUseCase = permissions.other_use_case?.trim();
    if (otherUseCase) {
      allowedUseCases.push(otherUseCase);
    }

    const response = await this.put<{ ok: boolean; data: ArtistTag }>(`/api/v1/artist/tags/${tagId}/permissions`, {
      ai_training_allowed: permissions.ai_training,
      allowed_use_cases: allowedUseCases,
      attribution_required: permissions.attribution,
      notes: permissions.notes || ''
    });
    return response.data;
  }

  async revokeTag(tagId: string): Promise<void> {
    await this.delete(`/api/v1/artist/tags/${tagId}`);
  }

  async getComplianceEvents(): Promise<ComplianceEvent[]> {
    type BackendEvent = {
      eventId: string;
      eventType: 'compliance' | 'similarity';
      scannedAt: string;
      outcome: 'allowed' | 'conditional' | 'restricted';
      reason?: string;
      companyName?: string | null;
      sourceName?: string;
      similarityFinding?: {
        topMatches?: Array<{ similarityPercent?: number }>;
      } | null;
    };

    const response = await this.get<{ ok: boolean; data: { count: number; events: BackendEvent[] } }>('/api/v1/artist/compliance-events');
    const events = response.data?.events || [];

    return events.map((event) => ({
      id: event.eventId,
      company_name: event.companyName || 'Unknown Company',
      date: event.scannedAt,
      outcome: event.outcome,
      artwork_name: event.sourceName || 'Unknown file',
      reason: event.reason || '',
      eventType: event.eventType,
      similarity_score: event.similarityFinding?.topMatches?.[0]?.similarityPercent
        ? event.similarityFinding.topMatches[0].similarityPercent / 100
        : undefined
    }));
  }

  // ==================== Company Endpoints ====================

  async checkCompanyAccess(): Promise<{ ok: boolean; data: Record<string, unknown> }> {
    return this.get('/api/v1/company/stub');
  }

  async getCompanyProfile(): Promise<CompanyProfile> {
    type BackendCompanyProfile = {
      companyName?: string;
      declaredUseCases?: string[];
      description?: string;
    };

    const response = await this.get<{ ok: boolean; data: { profile: BackendCompanyProfile | null } }>('/api/v1/company/profile');
    const profile = response.data?.profile;

    return {
      company_name: profile?.companyName || '',
      declared_use_cases: profile?.declaredUseCases || [],
      description: profile?.description || ''
    };
  }

  async updateCompanyProfile(data: CompanyProfileUpdate): Promise<CompanyProfile> {
    type BackendCompanyProfile = {
      companyName?: string;
      declaredUseCases?: string[];
      description?: string;
    };

    const response = await this.put<{ ok: boolean; data: { profile: BackendCompanyProfile | null } }>('/api/v1/company/profile', {
      company_name: data.company_name,
      declared_use_cases: data.declared_use_cases,
      description: data.description
    });
    const profile = response.data?.profile;

    return {
      company_name: profile?.companyName || '',
      declared_use_cases: profile?.declaredUseCases || [],
      description: profile?.description || ''
    };
  }

  async scanDataset(files: File[], zipFile?: File): Promise<ScanResponse> {
    type BackendScanReportItem = {
      sourceType: 'individual' | 'zip';
      sourceName: string;
      fileHash: string;
      securityTag?: string | null;
      reason?: string;
      agreementId?: string | null;
      agreementText?: string | null;
    };

    type BackendSimilarityFinding = {
      sourceType: 'individual' | 'zip';
      sourceName: string;
      fileHash: string;
      topMatches?: Array<{ securityTag?: string | null; similarityPercent?: number }>;
    };

    type BackendScanResponse = {
      summary: {
        totalFilesScanned: number;
        matchedItems: number;
        unmatchedFiles: number;
        allowed: number;
        conditional: number;
        restricted: number;
      };
      report: {
        allowed: BackendScanReportItem[];
        conditional: BackendScanReportItem[];
        restricted: BackendScanReportItem[];
        unmatched: BackendScanReportItem[];
      };
      similarityFindings?: {
        work_at_risk: BackendSimilarityFinding[];
        may_be_at_risk: BackendSimilarityFinding[];
      };
    };

    const formData = new FormData();
    files.forEach((file) => {
      formData.append('files', file);
    });
    if (zipFile) {
      formData.append('zip', zipFile);
    }

    const response = await this.request<{ ok: boolean; data: BackendScanResponse }>('POST', '/api/v1/company/scan', formData);
    const scanData = response.data;

    const mapReportItems = (items: BackendScanReportItem[]) =>
      items.map((item) => ({
        file_name: item.sourceName,
        tag_id: item.securityTag || undefined,
        reason: item.reason || undefined,
        agreement_id: item.agreementId || undefined,
        agreement_text: item.agreementText || undefined,
        source_type: item.sourceType,
        file_hash: item.fileHash
      }));

    const mapSimilarity = (findings: BackendSimilarityFinding[]) =>
      findings.map((finding) => ({
        file_name: finding.sourceName,
        artist_name: 'Protected Work',
        tag_id: finding.topMatches?.[0]?.securityTag || undefined,
        similarity_score: finding.topMatches?.[0]?.similarityPercent
          ? finding.topMatches[0].similarityPercent / 100
          : undefined,
        source_type: finding.sourceType,
        file_hash: finding.fileHash
      }));

    return {
      summary: scanData.summary,
      report: {
        allowed: mapReportItems(scanData.report.allowed || []),
        conditional: mapReportItems(scanData.report.conditional || []),
        restricted: mapReportItems(scanData.report.restricted || []),
        unmatched: mapReportItems(scanData.report.unmatched || [])
      },
      similarityFindings: scanData.similarityFindings
        ? {
            work_at_risk: mapSimilarity(scanData.similarityFindings.work_at_risk || []),
            may_be_at_risk: mapSimilarity(scanData.similarityFindings.may_be_at_risk || [])
          }
        : undefined
    };
  }

  async getScanHistory(): Promise<ScanResult[]> {
    const response = await this.get<{ ok: boolean; data: ScanResult[] }>('/api/v1/company/scan/history');
    return response.data || [];
  }
}

// ==================== Singleton Instance ====================

export const apiClient = new ApiClient(API_BASE_URL);

export default apiClient;
